### PR TITLE
docker-buildx 0.8.2 (new formula)

### PR DESCRIPTION
**Pull request by benmoss**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

I modeled this on a combination of https://github.com/Homebrew/homebrew-core/blob/master/Formula/buildkit.rb and https://github.com/Homebrew/homebrew-core/blob/master/Formula/docker-compose.rb

### CODE REVIEW

File: Formula/docker-buildx.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  url "https://github.com/docker/buildx.git",
+      tag:      "v0.8.2",
+      revision: "6224def4dd2c3d347eee19db595348c50d7cb491"
+  license "Apache-2.0"
+  head "https://github.com/docker/buildx.git", branch: "master"
+
+  livecheck do
+    url :stable
+    regex(/^v?(\d+(?:\.\d+)+)$/i)
+  end
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:13]: There's a default `livecheck` block that should be sufficient here
```suggestion
```

File: Formula/docker-buildx.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  bottle do
+    sha256 cellar: :any_skip_relocation, x86_64_linux: "c64de4f3c30f7a73ff9db637660c7aa0f00234368105b0a09fa8e24eebe910c3"
+    sha256 cellar: :any_skip_relocation, arm64_linux: "304d3d9822c75f98ad9cf57f0c234bcf326bbb96d791d551728cadd72a7a377f"
+    sha256 cellar: :any_skip_relocation, monterey: "95303b8b017d6805d35768244e66b41739745f81cb3677c0aefea231e484e227"
+  end
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:19]: we can remove this
```suggestion
```

File: Formula/docker-buildx.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    revision = Utils.git_head
+    ldflags = %W[
+      -s -w
+      -X github.com/docker/buildx/version.Version=#{version}
+      -X github.com/docker/buildx/version.Revision=#{revision}
+      -X github.com/docker/buildx/version.Package=github.com/docker/buildx
+    ]
+
+    system "go", "build", "-mod=vendor", "-trimpath",
+      "-ldflags", ldflags.join(" "), "-o", bin/"docker-buildx", "./cmd/buildx"
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:33]: we have a helper `std_go_args` that sets most of these by default:
```suggestion
    system "go", "build", *std_go_args(ldflags: ldflags), "./cmd/buildx"
```

File: Formula/docker-buildx.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  depends_on "go" => :build
+
+  def install
+    revision = Utils.git_head
+    ldflags = %W[
+      -s -w
+      -X github.com/docker/buildx/version.Version=#{version}
+      -X github.com/docker/buildx/version.Revision=#{revision}
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:28]: might as well inline this
```suggestion
      -X github.com/docker/buildx/version.Revision=#{Utils.git_head}
```

File: Formula/docker-buildx.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      -X github.com/docker/buildx/version.Package=github.com/docker/buildx
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:29]: this seems like the default, so we can omit it
```suggestion
```

File: Formula/docker-buildx.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    sha256 cellar: :any_skip_relocation, monterey: "95303b8b017d6805d35768244e66b41739745f81cb3677c0aefea231e484e227"
+  end
+
+  depends_on "go" => :build
+
+  def install
+    revision = Utils.git_head
+    ldflags = %W[
+      -s -w
+      -X github.com/docker/buildx/version.Version=#{version}
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:27]: The `docker-buildx` bundled with `Docker Desktop` has a leading `v`
```suggestion
      -X github.com/docker/buildx/version.Version=v#{version}
```

File: Formula/docker-buildx.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  def caveats
+    <<~EOS
+      docker-buildx is a Docker plugin. For Docker to find this plugin, symlink it:
+        mkdir -p ~/.docker/cli-plugins
+        ln -sfn #{opt_bin}/docker-buildx ~/.docker/cli-plugins/docker-buildx
+    EOS
+  end
+
+  test do
+    assert_match "github.com/docker/buildx #{version}", shell_output("#{bin}/docker-buildx version")
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:47]: could we add a more substantial test? for example, we could try to `docker-buildx build .` and check that it failed to connect to the docker daemon (see [whalebrew](https://github.com/Homebrew/homebrew-core/blob/328cec8f2bb3501d38b9154e94cc797f1abb1ffa/Formula/whalebrew.rb#L26-L27))

**Review by branchv — CHANGES_REQUESTED** [review]
thanks for your contribution, left a few suggestions

File: Formula/docker-buildx.rb (hunk 7 above)
Comment (benmoss) [reply Formula/docker-buildx.rb:47]: isn't that contingent on the environment that the test is running in? as in, wouldn't that fail if the environment did have a docker daemon running?

File: Formula/docker-buildx.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      docker-buildx is a Docker plugin. For Docker to find this plugin, symlink it:
+        mkdir -p ~/.docker/cli-plugins
+        ln -sfn #{opt_bin}/docker-buildx ~/.docker/cli-plugins/docker-buildx
+    EOS
+  end
+
+  test do
+    assert_match "github.com/docker/buildx v#{version}", shell_output("#{bin}/docker-buildx version")
+    output = shell_output(bin/"docker-buildx build . 2>&1", 1)
+    assert_match "Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?", output
```
Comment (branchvincent) [inline Formula/docker-buildx.rb:35]: I think we'll need to account for a different error on Linux, so this should work:
```suggestion
    assert_match(/(denied while trying to|Cannot) connect to the Docker daemon/, output)
```

**Review by chenrui333 — APPROVED** [review]

**Review by branchv — APPROVED** [review]
@benmoss thanks for your first contribution to Homebrew! 🍻

**Pull request closed without merging**